File: app/src/main/java/pl/trikimusic/controller/domain/model/SensorModels.kt
package pl.trikimusic.controller.domain.model

import kotlin.math.sqrt

data class Vector3(
    val x: Float,
    val y: Float,
    val z: Float,
) {
    val magnitude: Float
        get() = sqrt(x * x + y * y + z * z)

    operator fun minus(other: Vector3): Vector3 = Vector3(x - other.x, y - other.y, z - other.z)
}

data class TrikiSensorData(
    val frameIndex: Long,
    val timestampNanos: Long,
    val gyroscopeDps: Vector3,
    val accelerometerG: Vector3,
    val rawGyroscope: RawVector3,
    val rawAccelerometer: RawVector3,
    val status: Int,
) {
    val buttonPressed: Boolean
        get() = status and 0x01 != 0
}

data class RawVector3(
    val x: Short,
    val y: Short,
    val z: Short,
)

data class OrientationData(
    val pitch: Float = 0f,
    val roll: Float = 0f,
    val yaw: Float = 0f,
)

data class FilteredSensorData(
    val source: TrikiSensorData,
    val gyroscopeDps: Vector3,
    val accelerometerG: Vector3,
    val orientation: OrientationData,
) {
    val accelerationMagnitude: Float
        get() = accelerometerG.magnitude

    val gyroscopeMagnitude: Float
        get() = gyroscopeDps.magnitude
}

File: windows/TrikiMusicController.Windows/Models/SensorModels.cs
namespace TrikiMusicController_Windows.Models;

public readonly record struct Vector3f(float X, float Y, float Z)
{
    public float Magnitude => MathF.Sqrt(X * X + Y * Y + Z * Z);

    public static Vector3f operator -(Vector3f left, Vector3f right) =>
        new(left.X - right.X, left.Y - right.Y, left.Z - right.Z);
}

public readonly record struct RawVector3(short X, short Y, short Z);

public sealed record TrikiSensorData(
    long FrameIndex,
    long TimestampNanos,
    Vector3f GyroscopeDps,
    Vector3f AccelerometerG,
    RawVector3 RawGyroscope,
    RawVector3 RawAccelerometer,
    int Status);

public readonly record struct OrientationData(float Pitch, float Roll, float Yaw);

public sealed record FilteredSensorData(
    TrikiSensorData Source,
    Vector3f GyroscopeDps,
    Vector3f AccelerometerG,
    OrientationData Orientation)
{
    public float AccelerationMagnitude => AccelerometerG.Magnitude;
    public float GyroscopeMagnitude => GyroscopeDps.Magnitude;
}

public sealed record CalibrationProfile(
    float AccelerometerBiasX = 0,
    float AccelerometerBiasY = 0,
    float AccelerometerBiasZ = 0,
    float GyroscopeBiasX = 0,
    float GyroscopeBiasY = 0,
    float GyroscopeBiasZ = 0,
    float NeutralPitch = 0,
    float NeutralRoll = 0,
    float AccelerometerNoise = 0,
    float GyroscopeNoise = 0,
    int SampleCount = 0);
